### PR TITLE
Add sufficient deadlines and countermeasures to handle hung node scenario

### PR DESCRIPTION
**Pull request by shtripat**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Add sufficient deadlines and countermeasures to handle the hung node scenario

## Motivation and Context
This PR tries to address the hung node scenario by adding sufficient
deadlines and counter measures for such an eventuality.

## How to test this PR?
This PR already adds the relevant tests; you may reproduce them
locally as needed by following the mint automation piece. 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: .github/workflows/mint/nginx-4-node.conf [hunk 1]
```diff
@@ -23,10 +23,9 @@ http {
     # include /etc/nginx/conf.d/*.conf;
 
     upstream minio {
-        server minio1:9000;
-        server minio2:9000;
-        server minio3:9000;
-        server minio4:9000;
+        server minio1:9000 max_fails=1 fail_timeout=10s;
+        server minio2:9000 max_fails=1 fail_timeout=10s;
+        server minio3:9000 max_fails=1 fail_timeout=10s;
```
Comment (harshavardhana) [inline .github/workflows/mint/nginx-4-node.conf]: Removed minio4 since we are taking it down and pausing it. We add this to ensure that downed node is skipped however we test if other nodes respond properly.

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -899,6 +899,17 @@ func (er erasureObjects) getObjectFileInfo(ctx context.Context, bucket, object s
...
+
+		if totalResp >= minDisks && opts.FastGetObjInfo {
+			if countErrs(errs, errFileNotFound) >= minDisks || countErrs(errs, errFileVersionNotFound) >= minDisks {
+				err = errFileNotFound
+				if opts.VersionID != "" {
+					err = errFileVersionNotFound
+				}
+				break
+			}
+		}
```
Comment (harshavardhana) [inline cmd/erasure-object.go]: We verify if we have already seen an expected error and return right away.

File: internal/config/drive/drive.go [hunk 3]
```diff
@@ -65,8 +69,9 @@ func LookupConfig(kvs config.KVS) (cfg Config, err error) {
 	if err = config.CheckValidKeys(config.DriveSubSys, kvs, DefaultKVS); err != nil {
 		return cfg, err
 	}
+
 	// if not set. Get default value from environment
-	d := kvs.GetWithDefault(MaxTimeout, DefaultKVS)
+	d := env.Get(EnvMaxDriveTimeout, kvs.GetWithDefault(MaxTimeout, DefaultKVS))
```
Comment (harshavardhana) [inline internal/config/drive/drive.go]: found this bug which is missing we didn't honor MINIO_DRIVE_MAX_TIMEOUT by mistake

**Review by krisis — APPROVED** [review]

File: cmd/admin-server-info.go [hunk 4]
```diff
@@ -64,9 +64,11 @@ func getLocalServerProperty(endpointServerPools EndpointServerPools, r *http.Req
...
 					network[nodeName] = string(madmin.ItemOnline)
 				} else {
-					network[nodeName] = string(madmin.ItemOffline)
-					// log once the error
-					peersLogOnceIf(context.Background(), err, nodeName)
+					if xnet.IsNetworkOrHostDown(err, false) {
+						network[nodeName] = string(madmin.ItemOffline)
+					} else if xnet.IsNetworkOrHostDown(err, true) {
+						network[nodeName] = "connection attempt timedout"
+					}
```
Comment (harshavardhana) [inline cmd/admin-server-info.go:71]: To differentiate between actually "offline v/s timedout"

File: cmd/metrics-resource.go [hunk 5]
```diff
@@ -280,82 +281,70 @@ func collectDriveMetrics(m madmin.RealtimeMetrics) {
 func collectLocalResourceMetrics() {
 	var types madmin.MetricType = madmin.MetricsDisk | madmin.MetricNet | madmin.MetricsMem | madmin.MetricsCPU
```
Comment (harshavardhana) [inline cmd/metrics-resource.go]: https://github.com/minio/minio/pull/19762 PR sent here, its still part of this to allow testing.

File: cmd/server-main.go [hunk 6]
```diff
@@ -103,18 +103,6 @@ var ServerFlags = []cli.Flag{
...
-		Usage:  "custom connection READ deadline for incoming requests",
-		Hidden: true,
-		EnvVar: "MINIO_CONN_CLIENT_READ_DEADLINE",
-	},
-	cli.DurationFlag{
-		Name:   "conn-client-write-deadline",
-		Usage:  "custom connection WRITE deadline for outgoing requests",
-		Hidden: true,
-		EnvVar: "MINIO_CONN_CLIENT_WRITE_DEADLINE",
-	},
```
Comment (harshavardhana) [inline cmd/server-main.go]: not needed remove it.

File: internal/grid/connection.go [hunk 7]
```diff
@@ -1163,18 +1188,24 @@ func (c *Connection) handleMessages(ctx context.Context, conn net.Conn) {
...
 			return
 		}
 		_, err = buf.WriteTo(conn)
 		if err != nil {
-			gridLogIf(ctx, fmt.Errorf("ws write: %w", err))
+			if !xnet.IsNetworkOrHostDown(err, true) {
+				gridLogIf(ctx, fmt.Errorf("ws write: %w", err))
+			}
 			return
 		}
```
Comment (harshavardhana) [inline internal/grid/connection.go:1199]: Apply valid deadlines and only unexpected logs, not repeated network errors.

**Review by harshavardhana — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]

File: internal/grid/connection.go [hunk 8]
```diff
@@ -615,10 +616,6 @@ func (c *Connection) sendMsg(conn net.Conn, msg message, payload msgp.MarshalSiz
 	if c.outgoingBytes != nil {
 		c.outgoingBytes(int64(len(dst)))
 	}
-	err = conn.SetWriteDeadline(time.Now().Add(connWriteTimeout))
-	if err != nil {
-		return err
-	}
```
Comment (klauspost) [inline internal/grid/connection.go]: Why is this removed?

Comment (harshavardhana) [reply internal/grid/connection.go]: its not needed since the caller already provides the context @klauspost it can be retained I don't see any harm as such but it was just something that i thought is not needed.

let me add it back.
